src/BullOak.Repositories/Session/BaseRepoSession.cs

```diff
@@ -88,21 +87,31 @@ protected void Initialize(TState storedState, bool isNew)
             IsNewState = isNew;
         }
 
-        protected async Task PublishEvents(IHoldAllConfiguration configuration, object[] events, CancellationToken? cancellationToken = null)
+        public async Task<int> SaveChanges(DeliveryTargetGuarntee targetGuarantee =
+                DeliveryTargetGuarntee.AtLeastOnce,
+            CancellationToken? cancellationToken = null)
         {
-            for (var i = 0; i < events.Length; i++)
-            {
-                //We await each one to guarantee ordering of publishing, even though it would have been more performant
-                // to publish and await all of them with a WhenAll
-                await eventPublisher.Publish(events[i], cancellationToken);
-            }
+            var newEvents = NewEventsCollection.ToArray();
+            var sendEventsBeforeSaving = targetGuarantee == DeliveryTargetGuarntee.AtLeastOnce;
+
+            if (sendEventsBeforeSaving) await PublishEvents(configuration, newEvents, cancellationToken);
+            var retVal = await SaveChanges(newEvents, GetCurrentState(), cancellationToken);
+            if (!sendEventsBeforeSaving) await PublishEvents(configuration, newEvents, cancellationToken);
```

Comment:
Not too sure about this `sendEventsBeforeSaving` flag. 
Maybe it would be more clear to do it like this:
```C#
if (targetGuarantee == DeliveryTargetGuarntee.AtLeastOnce) await PublishEvents(configuration, newEvents, cancellationToken);
var retVal = await SaveChanges(newEvents, GetCurrentState(), cancellationToken);
if (targetGuarantee == DeliveryTargetGuarntee.AtMostOnce) await PublishEvents(configuration, newEvents, cancellationToken);
```

I am assuming that `DeliveryTargetGuarntee` is supposed to have only these two members, correct?
If I wanted to be more strict, I would add something like
```C#
switch (targetGuarantee)
{
  case DeliveryTargetGuarntee.AtLeastOnce: ... 
  case DeliveryTargetGuarntee.AtMostOnce:...
  default: throw Something.
}
```
but I guess this would be too much of a complication here.

Reply:
This was a tiny bit of optimization to save on comparing values twice and switch is not that performant either. This is not exposed to any developers and is not meant to be extended, so from my perspective is really fine. There are far less readable parts in the project anyway :) (see state emit).